### ISSUE TITLE
Vimscript: is this a design choice?

### ISSUE THREAD

**Issue by lane-core** [isovector/cornelis#144]
Hi, I've recently picked up Cornelis and I'm enamored thinking about the possibilities of how the vim ecosystem could assist Agda programming, as its a great foundation. 

I use neovim as my vim of choice, and am generally more fond of plugins written in Lua. I am curious as to whether the choice to code Cornelis in Vimscript is a design choice so as not to exclude vanilla vim users (which is totally legitimate), or whether the project would be open to Vimscript being converted to Lua, as it seems like a straightforward port and I imagine there are other developers who might prefer Lua as well (full disclosure: I have little desire to learn Vimscript, personally). Obviously, this means committing to the neovim ecosystem, but it feels like its gonna keep being a nice vibe there and I don't think that's something the project would regret.

I may try out a Lua port in a branch in my forked repo, so I wanted to inquire whether this is something you would be interested in as I'd then seek to contribute my results back to the project. Want to close off by saying I really love what you've done here @isovector and I think that there is a lot of low hanging fruit after the foundation you've laid that will be able to flesh out a lovely agda programming experience in vim.

**Comment by lane-core** [comment]
Sorry for taking so long to get back to this. I feel like this came off as if I was diminishing your work as yet another vimscirpt plugin, but the strong foundation I was referring to in my original post was the haskell core. I am way more agda-literate than I am in haskell at the moment, although that probably disposes me towards remedying that quickly. This is good because need to be living more in haskell world more in the near future, and I hope to contribute to the project in that way as I begin to feel fluent in haskell as Cornelis has replaced agda-mode for me and by then I might be familiar enough with how it works to see if I can attack some items in your issues list.

To the extent I was talking about the vimscript (or possibly lua) shell, it was to the extent that with relatively little addition to wrapper code we could probably improve the default experience quite a bit-- I had to refresh myself on some intermediate level vim config things just to get this working to what seemed like a sensible default (I can attach my config if you are curious what that means to me). Another motivation for wanting lua was that I want to do some experiments and see in what ways Cornelis could integrate into the wider neovim plugin ecosystem if one allowed access to these neovim apis. 

I will probably just port over the vimscript to lua to prototype possibilities there so I have concrete examples to point to down the line, but this general consideration was what originally prompted me to want to write the current ticket to inquire if the project is wedded to vimscript. Not only because I want to know whether I should prime my code with an eye towards submitting it somewhere (i.e. whether its just for my own fun and experiment), but also because regardless of the project's decision of the language I want to contribute nonetheless if I can, so if Cornelis is sticking with vimscript for the wrapper layer I will have to learn vimscript in order to make contributions on that front.

On the main, I hope this clarified my intention/motivations of why I wrote my original issue ticket. I want to extend my thanks again for your work, I've been delighted to use Cornelis and l look forward to helping out if possible in the future.

**Comment by isovector** [comment]
All contributions are welcome!

**Comment by 4e554c4c** [comment]
A lua port would be welcome. This plugin is already nvim only I believe. But as a maintainer of another RPC lua plugin there are things that are easier in vimscript, like finding the location of an executable relative to the source directory of your plugin.

**Comment by lane-core** [comment]
> If I were to give this a go, would I need to make the port backwards compatible? For example, would I still need to support configurations expressed via vimscript global variables?
> 
> Also, would it make sense to switch to [folke's which-key](https://github.com/folke/which-key.nvim), which is written in Lua, as opposed to the which-key plugin currently being used?
> 
> I am sure there's plenty to ask, but I cannot think of anything else right now.
> 
> EDIT: what about the project structure? `agda-input.vim` and `agda-matchpairs.vim` would probably end up in some folder.

I would be interested in picking this up with you -- also agree switching to which-key.nvim is ideal

**Comment by 4e554c4c** [comment]
Hey sorry, i forgot about this thread. I appreciate the enthusiasm, but maybe im feeling this a bit less than before.

First of all, I'm not to happy about replacing which-key with an alternative. I don't really think that a which-key plugin is the best use for agda-input completion. IMO which-key shows _keybindings_, while agda shortcuts are _completions_, which are slightly different. The latter should be able to be tab-completed, and should plug into a completion engine. This is why i wrote a plugin to a completion engine for Agda: https://github.com/4e554c4c/blink-cmp-agda-symbols
I also find folke personally difficult to work with, and I couldn't find any stability guarantees on the plugin, which would add a maintainership hazard should it ever change.

Next, re: variables and so on—this plugin is essentially 3 things: agda-input.vim (which i don't use), a haskell "core" plugin, and a small shim for finding/starting/communicating with the haskell bit which is mostly [nvim-hs.vim](https://github.com/neovimhaskell/nvim-hs.vim). The nvim-hs project takes care of a lot of the work here, especially process management and *finding the haskell executable in the hidden stack build directory* which I think would be a big pain to write from scratch in lua. From my [experience](https://github.com/4e554c4c/darkman.nvim/blob/main/lua/darkman/init.lua) with remote plugins, this is not the easiest thing to do; and maintaining two ways to communicate variables to haskell (vim variables and the lua `setup` idiom) would also be slightly difficult. I think backwards compatibility here is important, and maintaining it sounds hard.

tl;dr: the main part of the program is haskell and the vimscript part is maintained by someone else, which makes it easier on us.

That being said, if anyone has a good idea on how to transition the plugin to lua while maintaining backwards compatibility (for the most part) and not adding too much of a maintenance burden, im all ears.
